ansible_runner/utils.py

```diff
@@ -82,6 +83,25 @@ def check_isolation_executable_installed(isolation_executable):
         return False
 
 
+def stream_dir(directory, **additional):
+    buf = BytesIO()
+    with zipfile.ZipFile(buf, 'w', compression=zipfile.ZIP_DEFLATED, allowZip64=True) as archive:
+        if directory:
+            for dirpath, dirs, files in os.walk(directory):
+                relpath = os.path.relpath(dirpath, directory)
+                if relpath == ".":
+                    relpath = ""
+                for fname in files:
+                    archive.write(os.path.join(dirpath, fname), arcname=os.path.join(relpath, fname))
+
+        for fname, data in additional.items():
+            archive.writestr(fname, data)
+        archive.close()
+
+    payload = buf.getvalue()
+    return b'\n'.join((json.dumps({'zipfile': len(payload)}).encode('utf-8'), payload))
```

Comment:
Okay, I ran this against openshift origin, which is my large project repo example, it's 1.3 Gb, because my computer is not the best.

That returned bytes with a length of 868,468,620. It took 59 seconds to produce this.

That's roughly similar time to what I expect for doing a git clone... but the size in memory is more concerning. It would be nice if we could stream this from the transmitter to the worker and have the worker go straight to writing it out without it ever being fully held in memory.

I've been looking into the receptorctl options, it will take a zip file directly in `--payload`, but this requires writing it an additional time, which I also don't want to do, but could be worth the memory penalty.

Reply:
If you're running this via the Receptor remote worker (ie, you're submitting it as a unit of work to one Receptor node and running it on another), then Receptor itself will store the whole input payload on both nodes.  So at best, we would stream the zip content out of Runner to receptorctl, which streams it to the local Receptor node, which writes it to disk; then it is sent via Receptor to the remote Receptor node, which also writes it to disk; then it is streamed into the Runner worker, which decompresses it to disk.  So the minimum cost is one compressed and one uncompressed copy on each of the two nodes.

To avoid writing it to disk a second time when supplying it to receptorctl, you could do something along the lines of `runner transmit <args> | receptorctl <args> -f -`.

Reply:
When I did the AWX job directory isolation, we changed from using the project folder in-place to copying it into the job private data directory. This caused some complaints, because it meant we were copying a large folder. That ship has sailed. I accept that these changes will result in a new process that writes it out to disk 3 or 4 separate times (as opposed to 1, or 0 in ancient history).

My comment here was more directed at the memory use spike. I don't see any reason we need to have the entire project folder in memory, and doing so could cause problems. When I give the length of 868,468,620, that is the length of the object returned by this method.

Reply:
Are we talking about the returned artifacts dump?  Because I'm about to rip that out entirely.

Reply:
valid question, multiple things are calling `stream_dir`. I'm thinking of the case where `ansible-runner transmit` calls it for the whole private_data_dir. Within that directory, the `project/` folder is the concern.

Reply:
I'm afraid that's as intended.  How is the worker going to execute a job if it doesn't have a copy of the project?

Reply:
The concern isn't the size of `project/` on disk, the concern is loading that into memory - risking OOM. Here, the 2 calls are:

```python
self._output.write(utils.stream_dir(private_data_dir))
self._output.write(utils.stream_dir(artifact_dir))
```

Instead of returning this value, why not pass `self._output` as an argument instead of making a new `buf = BytesIO()`.

```diff
diff --git a/ansible_runner/streaming.py b/ansible_runner/streaming.py
index 2a1ff0c..0a373a6 100644
--- a/ansible_runner/streaming.py
+++ b/ansible_runner/streaming.py
@@ -49,7 +49,7 @@ class Transmitter(object):
         self._output.flush()
 
         private_data_dir = self.kwargs.get('private_data_dir')
-        self._output.write(utils.stream_dir(private_data_dir))
+        utils.stream_dir(private_data_dir, self._output)
         self._output.write(json.dumps({'eof': True}).encode('utf-8'))
         self._output.write(b'\n')
         self._output.flush()
@@ -120,7 +120,7 @@ class Worker(object):
         self._output.flush()
 
     def artifacts_handler(self, artifact_dir):
-        self._output.write(utils.stream_dir(artifact_dir))
+        utils.stream_dir(artifact_dir, self._output)
         self._output.flush()
 
     def finished_callback(self, runner_obj):
diff --git a/ansible_runner/utils.py b/ansible_runner/utils.py
index 71cbc38..1818a33 100644
--- a/ansible_runner/utils.py
+++ b/ansible_runner/utils.py
@@ -83,8 +83,8 @@ def check_isolation_executable_installed(isolation_executable):
         return False
 
 
-def stream_dir(directory):
-    buf = BytesIO()
+def stream_dir(directory, buf):
+    buff.write(b'\n'.join((json.dumps({'zipfile': len(payload)}).encode('utf-8'), b'')))
     with zipfile.ZipFile(buf, 'w', compression=zipfile.ZIP_DEFLATED, allowZip64=True) as archive:
         if directory:
             for dirpath, dirs, files in os.walk(directory):
@@ -95,9 +95,6 @@ def stream_dir(directory):
                     archive.write(os.path.join(dirpath, fname), arcname=os.path.join(relpath, fname))
         archive.close()
 
-    payload = buf.getvalue()
-    return b'\n'.join((json.dumps({'zipfile': len(payload)}).encode('utf-8'), payload))
-
 
 def dump_artifact(obj, path, filename=None):
     '''
```

Reply:
Oh, I see what you mean.  Good idea, if it can be made to work.

Reply:
There's no way to do that and still write the header object with the zip file size.

Reply:
That's true, I don't know what the python ZipFile stuff needs exactly.

As a last ditch, it could also work to send it to a temporary file.

Reply:
I'm pretty late coming back around to this, but filed an issue for this: https://github.com/ansible/ansible-runner/issues/604